### PR TITLE
Gap sellable quote integration

### PR DESCRIPTION
**Pull request by ruixuan-xu**

### CODE REVIEW

File: quote/gap.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	var policy models.Policy
+	err := json.Unmarshal(req, &policy)
+	lib.CheckError(err)
+
+	authToken, err := models.GetAuthTokenFromIdToken(r.Header.Get("Authorization"))
+	lib.CheckError(err)
+
+	Gap(authToken.Role, &policy)
+	policyJson, err := json.Marshal(policy)
+	return string(policyJson), nil, err
```
Comment (Yousef-Hammar) [inline quote/gap.go]: return also the policy

File: quote/gap.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+func getDataFrameFromCsv(path string) dataframe.DataFrame {
+	csvFile := lib.GetFilesByEnv("quote/gap_matrix_base.csv")
```
Comment (Yousef-Hammar) [inline quote/gap.go]: you are not using the path variable, so you will read always the same file

File: quote/gap.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+	return lib.CsvToDataframe(csvFile)
+}
+
+func Gap(role string, p *models.Policy) {
+	product, err := sellable.Gap(role, p)
+	lib.CheckError(err)
+
+	guarantees := make([]models.Guarante, 0, 10)
```
Comment (Yousef-Hammar) [inline quote/gap.go]: change with `make([]models.Guarante, 0)`

File: quote/gap.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+	netPrice := gapMultiplier * vehiclePrice
+	TaxOnPrice := netPrice * (tax / 100)
+	TaxOnPrice = lib.RoundFloat(TaxOnPrice, 2)
+	grossPrice := netPrice + TaxOnPrice
+
+	p.OffersPrices[offer] = map[string]*models.Price{
+		"singleInstallment": {
+			Net:      netPrice,
+			Tax:      TaxOnPrice,
+			Gross:    grossPrice,
```
Comment (Yousef-Hammar) [inline quote/gap.go]: round prices here

File: quote/gap.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+func initOfferPrices(
+	p *models.Policy,
+	offer string,
+	vehiclePrice float64,
+	gapMultiplier float64,
+	tax float64,
+) {
+	netPrice := gapMultiplier * vehiclePrice
+	TaxOnPrice := netPrice * (tax / 100)
```
Comment (Yousef-Hammar) [inline quote/gap.go]: first letter lower case

File: models/vehicle.go [hunk 6]
```diff
@@ -16,6 +17,8 @@ type Vehicle struct {
 	HasSatellite       bool      `firestore:"hasSatellite,omitempty"       json:"hasSatellite,omitempty"       bigquery:"-"` // PRESENZA SATELLITARE
 	IsFireTheftCovered bool      `firestore:"isFireTheftCovered,omitempty" json:"isFireTheftCovered,omitempty" bigquery:"-"` // COPERTURA FURTO/INCENDIO PREESISTENTE
 	MainUse            string    `firestore:"mainUse,omitempty"            json:"mainUse,omitempty"            bigquery:"-"` // UTILIZZO PRINCIPALE (privato/...)
+	IsElectric         bool      `firestore:"isElectric,omitempty"         json:"isElectric,omitempty"         bigquery:"-"` // SE E' UN VEICOLO ELETTRICO O NO
+	State              string    `firestore:"state,omitempty"              json:"state,omitempty"              bigquery:"-"` // VEICOLO NUOVO, USATO, ETC...
```
Comment (Yousef-Hammar) [inline models/vehicle.go]: change `State` into `Condition`

File: quote/gap.go [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+	policyJson, err := json.Marshal(policy)
+	return string(policyJson), nil, err
+}
+
+func getDataFrameFromCsv(path string) dataframe.DataFrame {
+	csvFile := lib.GetFilesByEnv("quote/gap_matrix_base.csv")
+	return lib.CsvToDataframe(csvFile)
+}
+
+func Gap(role string, p *models.Policy) {
```
Comment (Yousef-Hammar) [inline quote/gap.go]: GENERAL: do not use single letter names for variables (e.g., p -> policy/data)

File: quote/gap.go [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+	product, err := sellable.Gap(role, p)
+	lib.CheckError(err)
+
+	guarantees := make([]models.Guarante, 0, 10)
+	for _, g := range product.Companies[0].GuaranteesMap {
```
Comment (Yousef-Hammar) [inline quote/gap.go]: do not use single letter names for variable names

File: quote/gap.go [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+		guarantees = append(guarantees, *g)
+	}
+	p.Assets[0].Guarantees = guarantees
+
+	gapMatrices := map[string]dataframe.DataFrame{
+		"base":     getDataFrameFromCsv("quote/gap_matrix_base.csv"),
+		"complete": getDataFrameFromCsv("quote/gap_matrix_complete.csv"),
+	}
```
Comment (Yousef-Hammar) [inline quote/gap.go]: change into map[string]string where the key is the offerName and the value is the path to matrix file

File: quote/gap.go [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+
+func Gap(role string, p *models.Policy) {
+	product, err := sellable.Gap(role, p)
+	lib.CheckError(err)
+
+	guarantees := make([]models.Guarante, 0, 10)
+	for _, g := range product.Companies[0].GuaranteesMap {
+		guarantees = append(guarantees, *g)
+	}
+	p.Assets[0].Guarantees = guarantees
```
Comment (Yousef-Hammar) [inline quote/gap.go]: put into a separate function called `initGuarantees`

File: quote/gap.go [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+	tax := -1.0
+	for _, g := range product.Companies[0].GuaranteesMap {
+		tax = g.Tax
+		break
+	}
+	// Just in case
+	if tax == -1.0 {
+		lib.CheckError(errors.New("Error: no tax found"))
+	}
+
```
Comment (Yousef-Hammar) [inline quote/gap.go]: put this into a function named `getTaxValue`

File: quote/gap.go [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+	p.Assets[0].Guarantees = guarantees
+
+	gapMatrices := map[string]dataframe.DataFrame{
+		"base":     getDataFrameFromCsv("quote/gap_matrix_base.csv"),
+		"complete": getDataFrameFromCsv("quote/gap_matrix_complete.csv"),
+	}
+
+	provincesMatrix := getDataFrameFromCsv("enrich/provinces.csv")
+
+	residenceCode := p.Assets[0].Person.Residence.Locality
```
Comment (Yousef-Hammar) [inline quote/gap.go]: province is in `CityCode` field

File: quote/gap.go [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+
+func roundOffersPrices(offersPrices map[string]map[string]*models.Price) {
+	for offer, payments := range offersPrices {
+		for paymentType, price := range payments {
+			offersPrices[offer][paymentType].Net = lib.RoundFloat(price.Net, 2)
+			offersPrices[offer][paymentType].Tax = lib.RoundFloat(price.Tax, 2)
+			offersPrices[offer][paymentType].Gross = lib.RoundFloat(price.Gross, 2)
+		}
+	}
+}
```
Comment (Yousef-Hammar) [inline quote/gap.go]: remove this function and move the rounding to `calculateOfferPrices` function

File: quote/gap.go [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+	p.OffersPrices[offer] = map[string]*models.Price{
+		"singleInstallment": {
+			Net:      netPrice,
+			Tax:      TaxOnPrice,
+			Gross:    grossPrice,
+			Delta:    0.0,
+			Discount: 0.0,
+		},
+	}
+}
```
Comment (Yousef-Hammar) [inline quote/gap.go]: 1. change name into `calculateOfferPrices`
2. remove `vehiclePrice` input, you already have it inside policy object
3. remove `tax` input, you will call the function `getTaxValue`
4. remove `gapMultipliers` input, you will call the function `getMultipliers`

File: quote/gap.go [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+			break
+		}
+	}
+
+	taxString := strings.Replace(matrixAreaRow[duration], "%", "", -1)
+	tax, err := strconv.ParseFloat(taxString, 64)
+	lib.CheckError(err)
+
+	return tax / 100
+}
```
Comment (Yousef-Hammar) [inline quote/gap.go]: 1. change the name into `getMultipliers`
2. move here the map that contains csv paths

this function will take in input also the offerName and return the multipliers for that offer

File: sellable/sellable.go [hunk 16]
```diff
@@ -25,6 +26,12 @@ func Sellable(w http.ResponseWriter, r *http.Request) {
 				Method:  http.MethodPost,
 				Roles:   []string{models.UserRoleAll},
 			},
+			{
+				Route:   "/v1/sales/vehicle",
+				Handler: GapFx,
+				Method:  http.MethodPost,
+				Roles:   []string{models.UserRoleAll},
+			},
```
Comment (Yousef-Hammar) [inline sellable/sellable.go]: we don't need and enpoint for GAP

File: sellable/vehicle.go [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+		return "", models.Product{}, fmt.Errorf("cannot retrieve the product: %v", err)
+	}
+
+	jsonProduct, err := json.Marshal(product)
+	if err != nil {
+		return "", models.Product{}, fmt.Errorf("cannot generate the JSON: %v", err)
+	}
+
+	return string(jsonProduct), product, err
+}
```
Comment (Yousef-Hammar) [inline sellable/vehicle.go]: we don't need this, you can remove it

File: sellable/vehicle.go [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+
+	product, err := productForVehicle(p, role)
+	if err != nil {
+		return models.Product{}, fmt.Errorf("no products for this vehicle: %v", err)
+	}
+
+	return product, nil
+}
+
+func productForVehicle(p *models.Policy, r string) (models.Product, error) {
```
Comment (Yousef-Hammar) [inline sellable/vehicle.go]: change name into `getProduct`

**Review by Yousef-Hammar — CHANGES_REQUESTED** [review]

File: quote/gap.go [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+	Gap(authToken.Role, &policy)
+	policyJson, err := json.Marshal(policy)
+	return string(policyJson), policy, err
+}
+
+func Gap(role string, policy *models.Policy) {
+	product, err := sellable.Gap(role, policy)
+	lib.CheckError(err)
+
+	policy.Company = product.Companies[0].Name
```
Comment (Yousef-Hammar) [inline quote/gap.go]: not required

**Review by diogo-wopta — APPROVED** [review]

**Review by Yousef-Hammar — APPROVED** [review]

**Pull request merged**